include/deal.II/lac/petsc_snes.templates.h

```diff
@@ -0,0 +1,418 @@
+                       static_cast<PetscReal>(PETSC_DEFAULT);
+    PetscReal rtol = data.relative_tolerance > 0.0 ?
+                       data.relative_tolerance :
+                       static_cast<PetscReal>(PETSC_DEFAULT);
+    PetscReal stol = data.step_tolerance > 0.0 ?
+                       data.step_tolerance :
+                       static_cast<PetscReal>(PETSC_DEFAULT);
+
+    // Maximum number of iterations and function evaluations.
+    PetscInt maxit =
+      data.max_it >= 0 ? data.max_it : static_cast<PetscInt>(PETSC_DEFAULT);
+    PetscInt maxfe =
+      data.max_fe >= 0 ? data.max_fe : static_cast<PetscInt>(PETSC_DEFAULT);
+    AssertSNES(SNESSetTolerances(snes, atol, rtol, stol, maxit, maxfe));
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  void
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::reinit_matrix(
+    PMatrixType &P)
+  {
+    this->A = nullptr;
+    this->P = &P;
+  }
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  void
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::reinit_matrices(
+    AMatrixType &A,
+    PMatrixType &P)
+  {
+    this->A = &A;
+    this->P = &P;
+  }
+
+
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  unsigned int
+  NonlinearSolver<VectorType, PMatrixType, AMatrixType>::solve(VectorType &x)
+  {
+    auto snes_function_ =
+      [](SNES snes, Vec x, Vec f, void *ctx) -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      (void)snes;
+      auto user = static_cast<NonlinearSolver *>(ctx);
+
+      VectorType xdealii(x);
+      VectorType fdealii(f);
+      AssertUser(user->residual(xdealii, fdealii), "residual");
+      petsc_object_state_increase(f);
+      PetscFunctionReturn(0);
+    };
+
+    auto snes_jacobian_ =
+      [](SNES snes, Vec x, Mat A, Mat P, void *ctx) -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      (void)snes;
+      auto user = static_cast<NonlinearSolver *>(ctx);
+
+      VectorType  xdealii(x);
+      AMatrixType Adealii(A);
+      PMatrixType Pdealii(P);
+      AssertUser(user->jacobian(xdealii, Adealii, Pdealii), "jacobian");
+      petsc_object_state_increase(P);
+
+      // Handle the Jacobian-free case
+      // This call allows to resample the linearization point
+      // of the MFFD tangent operator
+      PetscBool flg;
+      AssertSNES(PetscObjectTypeCompare((PetscObject)A, MATMFFD, &flg));
+      if (flg)
+        {
+          AssertSNES(MatAssemblyBegin(A, MAT_FINAL_ASSEMBLY));
+          AssertSNES(MatAssemblyEnd(A, MAT_FINAL_ASSEMBLY));
+        }
+      else
+        petsc_object_state_increase(A);
+      PetscFunctionReturn(0);
+    };
+
+    auto snes_jacobian_with_setup_ =
+      [](SNES snes, Vec x, Mat A, Mat P, void *ctx) -> PetscErrorCode {
+      PetscFunctionBeginUser;
+      (void)snes;
+      auto user = static_cast<NonlinearSolver *>(ctx);
+
+      VectorType  xdealii(x);
```

Comment:
```suggestion
      VectorType  x_dealii(x);
```

Reply:
If you prefer this naming convention, I will update SNES and TS too. @bangerth your call

Reply:
I'll leave it this way then